programs/ssl/ssl_client2.c

```diff
@@ -2374,9 +2374,9 @@ int main( int argc, char *argv[] )
     }
     else
 #endif /* MBEDTLS_USE_PSA_CRYPTO */
-    if( ( ret = mbedtls_ssl_conf_psk( &conf, psk, psk_len,
+    if( ( psk_len > 0 ) && ( ( ret = mbedtls_ssl_conf_psk( &conf, psk, psk_len,
```

Comment:
This change is not correct. If `psk_len` is 0, `psk` and `psk_len` are silently ignored. There's no reason to do that. If you want to validate `psk_len`, do it and report an error if it's invalid. But in fact you don't need to validate it, since `mbedtls_ssl_conf_psk` does so. Just call `mbedtls_ssl_conf_psk`, i.e. no change here.

Same thing in `ssl_server2`.

Reply:
I am not sure with that. Programs ssl_client2 and ssl_server2 seems to be generic program tools which can be configured in different ways at the top of its c-files. One of that option is
#define DFL_PSK                 ""
and currently it means for me that this program don't use PSK. Checking correct parameters inside mbedtls_ssl_conf_psk is one think, but interpretation the configuration of this tools is something others in my opinion.
Are you sure that I should implement other solution privet returning MBEDTLS_ERR_SSL_BAD_INPUT_DATA from mbedtls_ssl_conf_psk?

Reply:
`DFL_xxx` are the DeFauLt values for command-line options. We never change the configuration of these tools by changing the values of `DFL_xxx`, we do it by passing command-line options. At this point in the code, `psk_len` is 0 if the program was called with no `psk=…` option or with `psk=` (setting `psk` to the empty string).

But I don't see how this is related to the code here. Some library functions have preconditions, and the test code should not call the library functions if these preconditions are not met. But `mbedtls_ssl_conf_psk` has well-defined behavior if `psk_len == 0`: it returns an error. So you don't need a check before calling it.

Reply:
Ah, I see why the change is needed. The code was brittle: it always called `mbedtls_ssl_conf_psk`, even when no PSK should have been configured. The library just ignores the PSK if it doesn't need one, so this was harmless. But now `mbedtls_ssl_conf_psk` returns an error if `psk` is empty (which is the default) and no PSK is desired.

It's reasonable to treat `psk` as conveying the PSK if it's non-empty, and conveying “don't use a PSK” if it's empty. (You can't do this with `psk_identity` because an empty PSK identity does make sense.) So I agree with your code change. Please update the documentation of `psk=%%s` above to add something like “No PSK if empty.”

Reply:
updated